fw/apm.c

```diff
@@ -930,52 +951,49 @@ tfw_apm_calc(TfwApmData *data)
  * tfw_apm_stats() should be used for calls in kernel context.
  * tfw_apm_stats_bh() should be used for calls in user context.
  */
-#define __tfw_apm_stats_body(apmdata, pstats, fn_lock, fn_unlock)	\
+#define __tfw_apm_stats_body(data, pstats, fn_lock, fn_unlock)		\
+({									\
 	unsigned int rdidx, seq = pstats->seq;				\
-	TfwApmData *data = apmdata;					\
 	TfwApmSEnt *asent;						\
 									\
-	BUG_ON(!apmdata);						\
-									\
 	smp_mb__before_atomic();					\
 	rdidx = atomic_read(&data->stats.rdidx);			\
 	asent = &data->stats.asent[rdidx % 2];				\
 									\
+	pstats->ith = tfw_pstats_ith;					\
 	fn_lock(&asent->rwlock);					\
 	memcpy(pstats->val, asent->pstats.val,				\
 	       T_PSZ * sizeof(pstats->val[0]));				\
 	fn_unlock(&asent->rwlock);					\
 	pstats->seq = rdidx;						\
 									\
-	return (seq != rdidx);
+	seq != rdidx;							\
+})
 
 int
-tfw_apm_stats_bh(void *apmdata, TfwPrcntlStats *pstats)
+tfw_apm_stats_bh(void *apmref, TfwPrcntlStats *pstats)
 {
-	__tfw_apm_stats_body(apmdata, pstats, read_lock_bh, read_unlock_bh);
+	TfwApmData *data;
+	BUG_ON(!apmref);
+	data = &((TfwApmRef *)apmref)->data;
+	return __tfw_apm_stats_body(data, pstats, read_lock_bh, read_unlock_bh);
 }
 
 int
-tfw_apm_stats(void *apmdata, TfwPrcntlStats *pstats)
+tfw_apm_stats(void *apmref, TfwPrcntlStats *pstats)
 {
-	__tfw_apm_stats_body(apmdata, pstats, read_lock, read_unlock);
+	TfwApmData *data;
+	BUG_ON(!apmref);
+	data = &((TfwApmRef *)apmref)->data;
+	return __tfw_apm_stats_body(data, pstats, read_lock, read_unlock);
 }
 
-/*
- * Verify that an APM Stats user using the same set of percentiles.
- *
- * Note: This module uses a single set of percentiles for all servers.
- * All APM Stats users must use the same set of percentiles.
- */
 int
-tfw_apm_pstats_verify(TfwPrcntlStats *pstats)
+tfw_apm_stats_global(TfwPrcntlStats *pstats)
 {
-	int i;
-
-	for (i = 0; i < T_PSZ; ++i)
-		if (pstats->ith[i] != tfw_pstats_ith[i])
-			return 1;
-	return 0;
+	BUG_ON(!tfw_apm_global_data);
+	return __tfw_apm_stats_body(tfw_apm_global_data, pstats, read_lock_bh,
+				    read_unlock_bh);
```

Comment:
Speaking of the introduced additional synchronization, it seems only locking in this place is introduces (besides the atomic operations for the cache hits & misses). The locking is happen between the timer and procfs reading, so it's not a huge concurrency.

However, timers work in the same softirq as the main networking code, so a huge tail latencies may occur due to a procfs reading. Maybe we should use seqlocks for explicit reader starvation to let softirqs always make progress.